### PR TITLE
Fix for generic constraint evaluation error in FSI

### PR DESCRIPTION
**Pull request by smoothdeveloper**
Add a test to reproduce #2411, I'll have a fix for review in separate commit

### CODE REVIEW

**Comment by smoothdeveloper** [comment]
@dotnet-bot test this please

**Comment by smoothdeveloper** [comment]
somehow my code doesn't compile with proto, interesting...

**Comment by smoothdeveloper** [comment]
I don't understand the failure now, the test I've added seems to pass but it then says `Compile Unexpectedly Failed: 256`.

I can run the same command locally and it works as I expect.

**Comment by smoothdeveloper** [comment]
Notes for reviewers:

* I'm using ` p.GetGenericTypeDefinition().IsAssignableFrom(a.GetGenericTypeDefinition())` to check the constraints, but I'm not sure it is most efficient nor totally correct (generic stuff is hard, wonder who was crazy enough to bring that to CLR...)
* I don't think this optimal, this will incur two calls to `convTypeRefAux` (first to resolve `mrefParameterTypes`, then the one we had before to resolve `argTs`), but I didn't find a way to handle this better, I'm not sure how bad is impact on performance

It does solve the issue I had, and it seem to not break any of our tests, although I don't know how extensive is FSI test coverage in the area of overload resolution / generic constraints.

Questions:
* what is the reason there is a different method overload resolution than FSC in this specific area?
* where is the equivalent code for FSC?
* why can't fsc-proto compile the code with my local functions if I don't add the type annotations (` error FS1103: Note: Lambda-lifting optimizations have not been applied because of the use of this local constrained generic function as a first class value. Adding type constraints may resolve this condition.`)?

File: tests/fsharpqa/Source/InteractiveSession/Misc/GenericConstraintWoes/issue2411/app.fsx [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#r "lib.dll"
+
+open System.Reflection
+
+type FooImpl() =
```
Comment (KevinRansom) [inline tests/fsharpqa/Source/InteractiveSession/Misc/GenericConstraintWoes/issue2411/app.fsx]: Foo triggers a failure on policheck which is one of our tools that validate code for offensive words,  please, let's not introduce any new occurrences of it.

Thanks

Kevin

File: src/absil/ilreflect.fs [hunk 2]
```diff
@@ -675,24 +674,58 @@ let queryableTypeGetMethodBySearch cenv emEnv parentT (mref:ILMethodRef) =
...
+        let satisfiesParameter (a: Type option) (p: Type) =
+            match a with
+            | None -> true
+            | Some a ->
+            if 
+                // obvious case
+                p.IsAssignableFrom a 
+            then true
+            elif
+                // both are generic
```
Comment (KevinRansom) [inline src/absil/ilreflect.fs]: unnecessary comment.

File: src/absil/ilreflect.fs [hunk 3]
```diff
@@ -675,24 +674,58 @@ let queryableTypeGetMethodBySearch cenv emEnv parentT (mref:ILMethodRef) =
         methInfo
     | _ ->
       (* Second, type match. Note type erased (non-generic) F# code would not type match but they have unique names *)
+
+        let satisfiesParameter (a: Type option) (p: Type) =
+            match a with
+            | None -> true
+            | Some a ->
+            if 
+                // obvious case
```
Comment (KevinRansom) [inline src/absil/ilreflect.fs]: unnecessary comment.

File: src/absil/ilreflect.fs [hunk 4]
```diff
@@ -675,24 +674,58 @@ let queryableTypeGetMethodBySearch cenv emEnv parentT (mref:ILMethodRef) =
         methInfo
     | _ ->
       (* Second, type match. Note type erased (non-generic) F# code would not type match but they have unique names *)
+
+        let satisfiesParameter (a: Type option) (p: Type) =
+            match a with
+            | None -> true
+            | Some a ->
+            if
```
Comment (KevinRansom) [inline src/absil/ilreflect.fs]: any chance this can be indented?

File: src/absil/ilreflect.fs [hunk 5]
```diff
@@ -675,24 +674,58 @@ let queryableTypeGetMethodBySearch cenv emEnv parentT (mref:ILMethodRef) =
...
+            if Array.length args <> Array.length ps then false
+            else Array.forall2 satisfiesParameter args ps
+       
         let select (methInfo:MethodInfo) =
             (* mref implied Types *)
             let mtyargTIs = getGenericArgumentsOfMethod methInfo 
+            
             if mtyargTIs.Length <> mref.GenericArity then false (* method generic arity mismatch *) else
+
+          (* methInfo implied Types *)
```
Comment (KevinRansom) [inline src/absil/ilreflect.fs]: align with code.

**Comment by smoothdeveloper** [comment]
@KevinRansom thanks for the reviews, I've made the final adjustments if you want to cherry pick that commit: https://github.com/smoothdeveloper/visualfsharp/commit/b622e129a2a348aa31cc7f2e53bff0d0ce64223c

@dsyme any clue about the questions I had: https://github.com/Microsoft/visualfsharp/pull/2590#issuecomment-285944009 ?

**Comment by dsyme** [comment]
> I'm using ``p.GetGenericTypeDefinition().IsAssignableFrom(a.GetGenericTypeDefinition())`` to check the constraints, but I'm not sure it is most efficient nor totally correct (generic stuff is hard, wonder who was crazy enough to bring that to CLR...)

Yes, this stuff is indeed quite hard.  It does feel like there may be additional corner cases lurking here but they will surely be very corner case.

**Pull request merged**